frame/staking/src/lib.rs

```diff
@@ -1418,12 +1579,72 @@ impl<T: Trait> Module<T> {
 		})
 	}
 
-	/// Select a new validator set from the assembled stakers and their role preferences.
+	/// Runs [`try_do_phragmen`] and updates the following storage items:
+	/// - [`Stakers`]
+	/// - [`SlotStake`]
+	/// - [`CurrentElected]
 	///
-	/// Returns the new `SlotStake` value and a set of newly selected _stash_ IDs.
+	/// If the election has been successful. It passes the new set upwards.
+	fn select_and_update_validators() -> Option<Vec<T::AccountId>> {
+		if let Some(ElectionResult::<T::AccountId, BalanceOf<T>> {
+			elected_stashes,
+			exposures,
+			era,
+			slot_stake,
+			compute,
+		}) = Self::try_do_phragmen() {
+			// Clear Stakers.
+			for v in Self::current_elected().iter() {
+				<Stakers<T>>::remove(v);
+			}
+
+			// Populate Stakers and write slot stake.
+			exposures.into_iter().for_each(|(s, e)| {
+				<Stakers<T>>::insert(s, e);
+			});
+
+			// Update slot stake.
+			<SlotStake<T>>::put(&slot_stake);
+
+			// Set the new validator set in sessions.
+			// Update current elected.
+			<CurrentElected<T>>::put(&elected_stashes);
+
+			// emit event
+			Self::deposit_event(RawEvent::StakingElection(compute));
+
+			Some(elected_stashes)
+		} else {
+			None
+		}
+	}
+
+	/// Select a new validator set from the assembled stakers and their role preferences. It tries
+	/// first to peek into [`QueuedElected`]. Otherwise, it runs a new phragmen.
 	///
-	/// Assumes storage is coherent with the declaration.
-	fn select_validators() -> (BalanceOf<T>, Option<Vec<T::AccountId>>) {
+	/// No storage item is updated.
+	fn try_do_phragmen() -> Option<ElectionResult<T::AccountId, BalanceOf<T>>> {
+		// a phragmen result from either a stored submission or locally executed one.
+		let somehow_phragmen_results = <QueuedElected<T>>::take().map(|offchain_result| {
+			debug_assert!(
+				Self::election_status() == OffchainElectionStatus::Received,
+				"offchain result exist but should not.",
+			);
+
+			offchain_result
+		}).or_else(|| Self::do_phragmen(ElectionCompute::OnChain));
+
+		// Either way, kill the flag. No more submitted solutions until further notice.
+		<ElectionStatus<T>>::kill();
+
+		somehow_phragmen_results
+	}
+
+	/// Execute phragmen and return the new results.
+	///
+	/// No storage item is updated.
+	fn do_phragmen(compute: ElectionCompute) -> Option<ElectionResult<T::AccountId, BalanceOf<T>>> {
```

Comment:
I guess off-chain we could do more rounds to get better results? That could be a parameter of the offchain worker.

Reply:
I made it as a todo up there. Either to be done here and in a follow up.